### PR TITLE
[Datasets] Implement push-based shuffle

### PR DESCRIPTION
**Pull request by stephanie-wang**
## Why are these changes needed?

The simple shuffle currently implemented in Datasets does not reliably scale past 1000+ partitions due to metadata and I/O overhead.

This PR adds an experimental shuffle implementation for a "push-based shuffle", as described in this [paper draft](https://arxiv.org/abs/2203.05072). This algorithm should see better performance at larger data scales. The algorithm works by merging intermediate map outputs at the reducer side while other map tasks are executing. Then, a final reduce task merges these merged outputs.

Currently, the PR exposes this option through the DatasetContext. It can also be set through a hidden OS environment variable (`RAY_DATASET_PUSH_BASED_SHUFFLE`). Once we have more comprehensive benchmarks, we can better document this option and allow the algorithm to be chosen at run time.

## Related issue number

Closes #23758.

### CODE REVIEW

**Comment by stephanie-wang** [comment]
@clarkzinzow @scv119 I'm going to run some some multi-node tests against this PR, but I could use some advice on how to test the algorithm internals in CI. The main things I want to check are that the merge stage includes the right number of tasks and that they are pipelined with the map stage tasks, but I'm not sure how to do that. Any tips?

**Comment by ericl** [comment]
Perhaps returning an info dict that contains stats you can check in a unit test?

**Comment by stephanie-wang** [comment]
> Perhaps returning an info dict that contains stats you can check in a unit test?

Yeah I was thinking the stats dict would be best for now, although I'm not sure yet that it includes the info I need. Btw, I think I'm doing something wrong for the merge stats because the number of blocks reported is much lower than expected. Could you take a quick look at that part?


> More generally, you could decompose the execute() operation into functions creating the topology and executing the topology of tasks. Then, it becomes easier to check the intermediate topology of tasks created.

Yeah... it's a bit complex at the moment because the map and merge tasks are interleaved, with `ray.wait` sync points in between. I'll keep the current version for now just to keep it simpler, but we could try this in the future with a more advanced lazy DAG representation.

File: python/ray/data/impl/shuffle.py
Comment (ericl) [inline python/ray/data/impl/shuffle.py:120]: Can you put this in a separate file?

File: python/ray/data/impl/shuffle.py
Comment (ericl) [inline python/ray/data/impl/shuffle.py:475]: This seems a bit magical. Can we have two separate shuffle ops instead?

File: python/ray/data/BUILD
Comment (ericl) [inline python/ray/data/BUILD:29]: Can you make the shuffle impl a DatasetContext option following the fusion optimization options? That will make it easier to write as a parameterized test without needing bazel env var hacks.

File: python/ray/data/impl/shuffle.py
Comment (stephanie-wang) [reply python/ray/data/impl/shuffle.py:366]: Makes sense, I'll just delete it right after submitting the merge tasks.

File: python/ray/data/impl/shuffle.py
Comment (stephanie-wang) [reply python/ray/data/impl/shuffle.py:252]: Yes, later we want to colocate the reducers with their upstream merge tasks, but this part is actually for the map-merge stage. We don't need to colocate map and merge tasks, but we do want to spread them evenly so we try to make sure there's always `merge_factor` map tasks : 1 merge task running on each node. That gives us groups of `merge_factor + 1` tasks that we need to spread.

I'll add this as a comment!

File: python/ray/data/impl/shuffle.py
Comment (stephanie-wang) [reply python/ray/data/impl/shuffle.py:286]: Yeah I think the ideal case would be to always determine automatically. I'll add a TODO about the merge factor < 1, though, I don't think this is currently supported.

**Comment by stephanie-wang** [comment]
An initial run of `sort` on 1TB, 1k partitions produces 1h on 20 nodes vs 1h20min with simple shuffle. Looking into the performance a bit, there is definitely room for improvement on this PR, specifically around improving pipelining during the map-merge stage. But I think we should merge this version in first just to get the skeleton in there.

File: python/ray/data/impl/push_based_shuffle.py
Comment (jjyao) [inline python/ray/data/impl/push_based_shuffle.py:175]: Should we also use NodeAffinitySchedulingStrategy to colocate `merge_factor` number of map tasks with the merge task?

Comment (stephanie-wang) [reply python/ray/data/impl/push_based_shuffle.py:175]: I think it is actually better to let Ray schedule the map tasks in case there are other factors like data locality being used to schedule those. I'll make a note of this in the code!

File: python/ray/data/impl/push_based_shuffle.py
Comment (ericl) [inline python/ray/data/impl/push_based_shuffle.py:180]: Nit: It would be nicer for readability to put execute() at the top of the class, and the helper functions lower.

File: python/ray/data/impl/push_based_shuffle.py
Comment (ericl) [inline python/ray/data/impl/push_based_shuffle.py:242]: Could we unit test the above topology configuration logic? Concretely, we could split it out into helper method that returns a struct/dict, and we can unit test the topology configured for various cluster scenarios. This would also help with splitting the code up for readability.

File: python/ray/data/impl/push_based_shuffle.py
Comment (ericl) [inline python/ray/data/impl/push_based_shuffle.py:268]: I don't know if this is easy to do, but ideally we'd pull out the pipelining / round logic into a generic method. That way, the map task logic etc. isn't mixed in with progress bars and the pipelining.

File: python/ray/data/impl/push_based_shuffle.py
Comment (ericl) [inline python/ray/data/impl/push_based_shuffle.py:386]: Add some test to test_stats.py?

File: python/ray/data/impl/push_based_shuffle.py
Comment (scv119) [inline python/ray/data/impl/push_based_shuffle.py:117]: nit: type hints help readability!

File: python/ray/data/impl/push_based_shuffle.py
Comment (scv119) [inline python/ray/data/impl/push_based_shuffle.py:225]: nit: num_cpus_per_node_map

File: python/ray/data/impl/push_based_shuffle.py
Comment (scv119) [reply python/ray/data/impl/push_based_shuffle.py:242]: +1

File: python/ray/data/impl/push_based_shuffle.py
Comment (scv119) [inline python/ray/data/impl/push_based_shuffle.py:269]: can we wrap this block into a helper function as well, for readability? e.g., `execute_map_stage`

File: python/ray/data/impl/push_based_shuffle.py
Comment (scv119) [inline python/ray/data/impl/push_based_shuffle.py:342]: similarly, this could belong to a reducer helper function if possible

**Review by scv119 — COMMENTED** [review]
This looks great! Some high level code structure suggestions.

File: python/ray/data/context.py
Comment (stephanie-wang) [reply python/ray/data/context.py:33]: Ah right, good point... It only needs to be set on whoever calls the Dataset shuffle call; after that I think the context will get propagated through the usual mechanism. Do you think that's an issue?

**Review by ericl — APPROVED** [review]

**Comment by ericl** [comment]
I think some tests are failing here. I reviewed this at a high level and am ok with merging this to unblock further testing, though we might want to revisit the code structure at a later point (could be after we're happy with the performance).

**Review by scv119 — APPROVED** [review]

**Review by clarkzinzow — APPROVED** [review]
LGTM, same thoughts as @ericl: I have some ideas of how to improve the code structure here, but this looks good enough to merge, so let's unblock next steps and revisit after perf testing.

File: python/ray/data/impl/push_based_shuffle.py
Comment (franklsf95) [inline python/ray/data/impl/push_based_shuffle.py:120]: I don't understand what fractional merge factors mean. Like, how would a merge task merge 1.5x map outputs? And why is this related with reduce/map workload ratio?

**Review by frank-lsf — COMMENTED** [review]
Awesome seeing this work going into production! I left a few questions inline.

One thing we should keep in mind is data skew. The current pipelining schedule will not work very well if the input partitions are too imbalanced.

File: python/ray/data/impl/push_based_shuffle.py
Comment (stephanie-wang) [reply python/ray/data/impl/push_based_shuffle.py:120]: It just means you have more merge tasks than map tasks. So if you have map:merge = 1/2 and 10 map tasks for every round, each merge task will output 1/2 of a partition's worth of data. Presumably you'd also want to have many more reduce tasks than map tasks in this scenario.

But yeah, there's no evidence yet that we'd want to support this case. I haven't seen a case yet where the reduce-side computation is heavier than the map.

**Pull request merged**
